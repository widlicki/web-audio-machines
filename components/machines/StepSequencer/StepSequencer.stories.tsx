import { ComponentMeta, ComponentStory } from '@storybook/react'
import StepSequencer from './StepSequencer'

export default {
  title: 'components/machines/StepSequencer',
  component: StepSequencer,
  parameters: {
    layout: 'centered',
  },
} as ComponentMeta<typeof StepSequencer>

const Template: ComponentStory<typeof StepSequencer> = (args) => (
  <StepSequencer {...args} />
)

export const Default = Template.bind({})
Default.args = {}
