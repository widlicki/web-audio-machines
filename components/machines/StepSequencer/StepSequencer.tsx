import { Grid } from '@mui/material'
import StepPad from '../StepPad/StepPad'

const StepSequencer = () => {
  return (
    <Grid container direction="row" justifyContent="center" alignItems="center">
      <Grid container direction="row" xs={6}>
        <Grid item xs={3}>
          <StepPad active={false} enabled={false} on={false} />
        </Grid>
        <Grid item xs={3}>
          <StepPad active={false} enabled={false} on={false} />
        </Grid>
        <Grid item xs={3}>
          <StepPad active={false} enabled={false} on={false} />
        </Grid>
        <Grid item xs={3}>
          <StepPad active={false} enabled={false} on={false} />
        </Grid>
      </Grid>
      <Grid container direction="row" xs={6}>
        <Grid item xs={3}>
          <StepPad active={false} enabled={false} on={false} />
        </Grid>
        <Grid item xs={3}>
          <StepPad active={false} enabled={false} on={false} />
        </Grid>
        <Grid item xs={3}>
          <StepPad active={false} enabled={false} on={false} />
        </Grid>
        <Grid item xs={3}>
          <StepPad active={false} enabled={false} on={false} />
        </Grid>
      </Grid>
    </Grid>
  )
}

export default StepSequencer
