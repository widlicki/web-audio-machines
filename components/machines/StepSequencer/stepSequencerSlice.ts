/* eslint-disable no-param-reassign */
import { createSlice } from '@reduxjs/toolkit'
import type { AppState } from '../../../store/index'

export interface StepSequencerState {
  maxNumSteps: 16 | 32 | 64
  minNumSteps: number
  numSteps: number
}

const initialState: StepSequencerState = {
  maxNumSteps: 32,
  minNumSteps: 8,
  numSteps: 16,
}

export const stepSequencerSlice = createSlice({
  name: 'stepSequencer',
  initialState,
  reducers: {
    decNumSteps: (state) => {
      state.numSteps =
        state.numSteps >= state.minNumSteps
          ? state.numSteps - 1
          : state.numSteps
    },
    incNumSteps: (state) => {
      state.numSteps =
        state.numSteps <= state.maxNumSteps
          ? state.numSteps + 1
          : state.numSteps
    },
  },
})

export const { incNumSteps, decNumSteps } = stepSequencerSlice.actions

export const selectNumSteps = (state: AppState) => state.stepSequencer.numSteps
export const selectMaxNumSteps = (state: AppState) =>
  state.stepSequencer.maxNumSteps
export const selectMinNumSteps = (state: AppState) =>
  state.stepSequencer.minNumSteps

export default stepSequencerSlice.reducer
