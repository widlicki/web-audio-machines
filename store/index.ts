import { configureStore } from '@reduxjs/toolkit'
import playbackControlReducer from '../components/machines/PlaybackControl/playbackControlSlice'
import stepControlReducer from '../components/machines/StepControl/stepControlSlice'

export function makeStore() {
  return configureStore({
    reducer: {
      playbackControl: playbackControlReducer,
      stepControl: stepControlReducer,
    },
  })
}

const store = makeStore()

export default store

export type AppState = ReturnType<typeof store.getState>

export type AppDispatch = typeof store.dispatch
