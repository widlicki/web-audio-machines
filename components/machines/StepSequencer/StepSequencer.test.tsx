import { render } from '@testing-library/react'
import { Provider } from 'react-redux'
import { makeStore } from '../../../store'
import StepSequencer from './StepSequencer'
import { Default } from './StepSequencer.stories'

let component: any
let store: any

const createConnectedComponent = (args) => {
  return (
    <Provider store={store}>
      <StepSequencer {...args} />
    </Provider>
  )
}

describe('Step Sequencer Stories', () => {
  beforeEach(() => {
    store = makeStore()
  })

  new Map([['default', Default]]).forEach(({ args }, key) => {
    it(`renders ${key}`, () => {
      component = createConnectedComponent(args)
      const { container } = render(component)
      expect(container).toMatchSnapshot()
    })
  })
})
