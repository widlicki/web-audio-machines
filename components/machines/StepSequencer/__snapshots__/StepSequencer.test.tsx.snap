// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Step Sequencer Stories renders default 1`] = `
<div>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-7mywch-MuiPaper-root-MuiCard-root"
  >
    <div
      class="MuiCardHeader-root css-1aseecf-MuiCardHeader-root"
    >
      <div
        class="MuiCardHeader-content css-1qbkelo-MuiCardHeader-content"
      >
        <span
          class="MuiTypography-root MuiTypography-subtitle2 MuiCardHeader-title css-1sr7afs-MuiTypography-root"
        >
          Step Sequencer
        </span>
      </div>
    </div>
    <div
      class="MuiCardContent-root css-198pza3-MuiCardContent-root"
    >
      <div
        class="MuiGrid-root MuiGrid-container css-1hbmzt3-MuiGrid-root"
      >
        <div
          class="MuiGrid-root MuiGrid-container MuiGrid-grid-xs-6 css-dzaaci-MuiGrid-root"
        >
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-direction-xs-column MuiGrid-grid-xs-3 css-10jz4s5-MuiGrid-root"
          >
            <div
              class="MuiBox-root css-zf09ms"
            />
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-direction-xs-column MuiGrid-grid-xs-3 css-10jz4s5-MuiGrid-root"
          >
            <div
              class="MuiBox-root css-zf09ms"
            />
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-direction-xs-column MuiGrid-grid-xs-3 css-10jz4s5-MuiGrid-root"
          >
            <div
              class="MuiBox-root css-zf09ms"
            />
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-direction-xs-column MuiGrid-grid-xs-3 css-10jz4s5-MuiGrid-root"
          >
            <div
              class="MuiBox-root css-zf09ms"
            />
          </div>
        </div>
        <div
          class="MuiGrid-root MuiGrid-container MuiGrid-grid-xs-6 css-dzaaci-MuiGrid-root"
        >
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-direction-xs-column MuiGrid-grid-xs-3 css-10jz4s5-MuiGrid-root"
          >
            <div
              class="MuiBox-root css-zf09ms"
            />
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-direction-xs-column MuiGrid-grid-xs-3 css-10jz4s5-MuiGrid-root"
          >
            <div
              class="MuiBox-root css-zf09ms"
            />
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-direction-xs-column MuiGrid-grid-xs-3 css-10jz4s5-MuiGrid-root"
          >
            <div
              class="MuiBox-root css-zf09ms"
            />
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-direction-xs-column MuiGrid-grid-xs-3 css-10jz4s5-MuiGrid-root"
          >
            <div
              class="MuiBox-root css-zf09ms"
            />
          </div>
        </div>
        <div
          class="MuiGrid-root MuiGrid-container MuiGrid-grid-xs-6 css-dzaaci-MuiGrid-root"
        >
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-direction-xs-column MuiGrid-grid-xs-3 css-10jz4s5-MuiGrid-root"
          >
            <div
              class="MuiBox-root css-zf09ms"
            />
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-direction-xs-column MuiGrid-grid-xs-3 css-10jz4s5-MuiGrid-root"
          >
            <div
              class="MuiBox-root css-zf09ms"
            />
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-direction-xs-column MuiGrid-grid-xs-3 css-10jz4s5-MuiGrid-root"
          >
            <div
              class="MuiBox-root css-zf09ms"
            />
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-direction-xs-column MuiGrid-grid-xs-3 css-10jz4s5-MuiGrid-root"
          >
            <div
              class="MuiBox-root css-zf09ms"
            />
          </div>
        </div>
        <div
          class="MuiGrid-root MuiGrid-container MuiGrid-grid-xs-6 css-dzaaci-MuiGrid-root"
        >
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-direction-xs-column MuiGrid-grid-xs-3 css-10jz4s5-MuiGrid-root"
          >
            <div
              class="MuiBox-root css-zf09ms"
            />
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-direction-xs-column MuiGrid-grid-xs-3 css-10jz4s5-MuiGrid-root"
          >
            <div
              class="MuiBox-root css-zf09ms"
            />
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-direction-xs-column MuiGrid-grid-xs-3 css-10jz4s5-MuiGrid-root"
          >
            <div
              class="MuiBox-root css-zf09ms"
            />
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-direction-xs-column MuiGrid-grid-xs-3 css-10jz4s5-MuiGrid-root"
          >
            <div
              class="MuiBox-root css-zf09ms"
            />
          </div>
        </div>
        <div
          class="MuiGrid-root MuiGrid-container MuiGrid-grid-xs-6 css-dzaaci-MuiGrid-root"
        >
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-direction-xs-column MuiGrid-grid-xs-3 css-10jz4s5-MuiGrid-root"
          >
            <div
              class="MuiBox-root css-zf09ms"
            />
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-direction-xs-column MuiGrid-grid-xs-3 css-10jz4s5-MuiGrid-root"
          >
            <div
              class="MuiBox-root css-zf09ms"
            />
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-direction-xs-column MuiGrid-grid-xs-3 css-10jz4s5-MuiGrid-root"
          >
            <div
              class="MuiBox-root css-zf09ms"
            />
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-direction-xs-column MuiGrid-grid-xs-3 css-10jz4s5-MuiGrid-root"
          >
            <div
              class="MuiBox-root css-zf09ms"
            />
          </div>
        </div>
        <div
          class="MuiGrid-root MuiGrid-container MuiGrid-grid-xs-6 css-dzaaci-MuiGrid-root"
        >
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-direction-xs-column MuiGrid-grid-xs-3 css-10jz4s5-MuiGrid-root"
          >
            <div
              class="MuiBox-root css-zf09ms"
            />
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-direction-xs-column MuiGrid-grid-xs-3 css-10jz4s5-MuiGrid-root"
          >
            <div
              class="MuiBox-root css-zf09ms"
            />
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-direction-xs-column MuiGrid-grid-xs-3 css-10jz4s5-MuiGrid-root"
          >
            <div
              class="MuiBox-root css-zf09ms"
            />
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-direction-xs-column MuiGrid-grid-xs-3 css-10jz4s5-MuiGrid-root"
          >
            <div
              class="MuiBox-root css-zf09ms"
            />
          </div>
        </div>
        <div
          class="MuiGrid-root MuiGrid-container MuiGrid-grid-xs-6 css-dzaaci-MuiGrid-root"
        >
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-direction-xs-column MuiGrid-grid-xs-3 css-10jz4s5-MuiGrid-root"
          >
            <div
              class="MuiBox-root css-zf09ms"
            />
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-direction-xs-column MuiGrid-grid-xs-3 css-10jz4s5-MuiGrid-root"
          >
            <div
              class="MuiBox-root css-zf09ms"
            />
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-direction-xs-column MuiGrid-grid-xs-3 css-10jz4s5-MuiGrid-root"
          >
            <div
              class="MuiBox-root css-zf09ms"
            />
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-direction-xs-column MuiGrid-grid-xs-3 css-10jz4s5-MuiGrid-root"
          >
            <div
              class="MuiBox-root css-zf09ms"
            />
          </div>
        </div>
        <div
          class="MuiGrid-root MuiGrid-container MuiGrid-grid-xs-6 css-dzaaci-MuiGrid-root"
        >
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-direction-xs-column MuiGrid-grid-xs-3 css-10jz4s5-MuiGrid-root"
          >
            <div
              class="MuiBox-root css-zf09ms"
            />
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-direction-xs-column MuiGrid-grid-xs-3 css-10jz4s5-MuiGrid-root"
          >
            <div
              class="MuiBox-root css-zf09ms"
            />
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-direction-xs-column MuiGrid-grid-xs-3 css-10jz4s5-MuiGrid-root"
          >
            <div
              class="MuiBox-root css-zf09ms"
            />
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-direction-xs-column MuiGrid-grid-xs-3 css-10jz4s5-MuiGrid-root"
          >
            <div
              class="MuiBox-root css-zf09ms"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
